utils/src/vector.c

```diff
@@ -0,0 +1,152 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "vector.h"
+
+/*!
+ * @brief	Finds the smallest power of two greater than a given number
+ *
+ * @param	base The base number
+ *
+ * @return  The power of two found in the function
+ */
+static size_t roundup(size_t base) {
+    if (base == 0 || base > SIZE_MAX / 2) {
+        return 0;
+    }
+
+    size_t acc = 1;
+    while (base > acc) {
+        acc *= 2;
+    }
+    return acc;
+}
+
+/*!
+ * @brief    Initializes a new vector of a certain size
+ *
+ * @param    item_size Size of the items to be stored in the vector in bytes,
+ *           must be greater than 0
+ *
+ * @param    init_size Initial size of vector in number of items
+ *
+ * @return   A pointer to the new vector, NULL when item_size is zero or memory
+ *           allocation fails
+ */
+vector_t* init_vector_of_size(size_t item_size, size_t init_size) {
+    if (item_size == 0) {
+        return NULL;
+    }
+
+    vector_t* vec = malloc(sizeof(vector_t));
+    if (vec == NULL) {
+      return NULL;
+    }
+
+    vec->item_size = item_size;
+    vec->used = 0;
+    size_t true_init_size = roundup(init_size);
+    vec->capacity = true_init_size;
+
+    /* 
+     * If they requested 0 bytes, no need to malloc any memory 
+     * Realloc will act like malloc if the given pointer is null
+     */
+    if (init_size > 0) {
+        vec->contents = malloc(sizeof(char) * true_init_size * item_size);
+
+        /* Check for memory allocation failure */
+        if (vec->contents == NULL) {
+            return NULL;
```

Comment:
You need to free vec, if this turns out to be true, else there will be a memory leak

Reply:
Good catch!